### PR TITLE
feat: Data room landing page

### LINKED ISSUE #354: [EPIC] Data room landing page
## Subtasks
- [x] #376 
- [x] #355 
- [x] #377
- [x] #378
- [x] #379
- [x] #380
- [x] #381
- [x] #382
- [x] #383
- [x] #384
- [x] #385

### PR DESCRIPTION
**Pull request by DiogoSoaress**
Related to #354 

Implements the sections 
#377 - Percentage of USDC stored
#378 - CryptoPunks stored
#379 - Total volume transferred 
#380 - Transactions on chain

### CODE REVIEW

File: src/components/DataRoom/CryptoPunks/index.tsx [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+import type { ReactNode } from 'react'
+import type { MotionValue } from 'framer-motion'
+import css from './styles.module.css'
+import LinksWrapper from '../LinksWrapper'
+import { getColor } from './utils'
+import CryptoPunk from '@/public/images/DataRoom/cryptopunk-silhouette.svg'
+import { useIsMediumScreen } from '@/hooks/useMaxWidth'
+
+const CRYPTOPUNKS_PERCENTAGE = '14%'
+const CRYPTOPUNKS_FRACTION = '1369/10,000'
```
Comment (DiogoSoaress) [inline src/components/DataRoom/CryptoPunks/index.tsx:14]: This hardcoded values will be replaced in another PR

File: src/components/DataRoom/Hero/index.tsx [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+import type { BaseBlock } from '@/components/Home/types'
+
+const Hero = ({ title, text }: BaseBlock) => (
+  <Container>
+    <Box height="100vh" display="flex" flexDirection="column" alignItems="center" justifyContent="center">
+      <Typography variant="h1">{title}</Typography>
+      <Typography color="primary.light">{text}</Typography>
+    </Box>
+  </Container>
+)
```
Comment (DiogoSoaress) [inline src/components/DataRoom/Hero/index.tsx:11]: This is a placeholder and will be replaced in a different PR. It's included in the PR so the user can experience scrolling into the next section.

File: src/components/DataRoom/LinksWrapper/index.tsx [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+import css from './styles.module.css'
+
+// TODO: define the content to display in the Socials component buttons
+const LinksWrapper = ({ title, href }: Link) => {
+  return (
+    <div className={css.wrapper}>
+      <SafeLink href={href}>
+        <LinkButton>{title}</LinkButton>
+      </SafeLink>
+      <Socials authors={[]} title="" />
```
Comment (DiogoSoaress) [inline src/components/DataRoom/LinksWrapper/index.tsx:14]: This content will be addressed in a different PR

File: src/components/DataRoom/CryptoPunks/index.tsx [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+import { useScroll, motion, useTransform } from 'framer-motion'
+import type { ReactNode } from 'react'
+import type { MotionValue } from 'framer-motion'
+import css from './styles.module.css'
+import LinksWrapper from '../LinksWrapper'
+import { getRandomColor } from '@/components/DataRoom/CryptoPunks/utils/getRandomColor'
+import CryptoPunk from '@/public/images/DataRoom/cryptopunk-silhouette.svg'
+import { useIsMediumScreen } from '@/hooks/useMaxWidth'
+
+const CRYPTOPUNKS_PERCENTAGE = '14%'
```
Comment (iamacook) [inline src/components/DataRoom/CryptoPunks/index.tsx:13]: You are calculating percentages elsewhere. I'd suggest we remain consistent and include the non-percentage value, calculating it.

File: src/components/DataRoom/CryptoPunks/index.tsx [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+const CRYPTOPUNKS_FRACTION = '1369/10,000'
```
Comment (iamacook) [inline src/components/DataRoom/CryptoPunks/index.tsx:14]: By hardcoding this, it will not format correctly. As before, I'd store the value _as is_ and format it accordingly.

File: package.json [hunk 6]
```diff
@@ -31,6 +31,7 @@
     "fuse.js": "^6.6.2",
     "isomorphic-dompurify": "^1.0.0",
     "lodash": "^4.17.21",
+    "matter-js": "^0.20.0",
```
Comment (iamacook) [inline package.json:34]: Why do we need to install this when we have framer?

File: src/components/DataRoom/CryptoPunks/utils.ts [hunk 7]
```diff
@@ -0,0 +1,9 @@
+export function getColor() {
+  const chance = Math.random()
+
+  if (chance <= 0.24) return '#12FF80' // Green (14% CHANCE)
+  if (chance <= 0.17) return '#12A154' // Dark Green (3% CHANCE)
+  if (chance <= 0.2) return '#124228' // Darker Green (3% CHANCE)
+
+  return 'currentColor'
```
Comment (iamacook) [inline src/components/DataRoom/CryptoPunks/utils.ts:8]: Why do we include chance in this? It would be a lot simpler to just have an array and pick a random index.

File: src/components/DataRoom/CryptoPunks/utils/getRandomColor.ts [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+ */
+export function getRandomColor() {
+  const chance = Math.random()
+
+  if (chance <= 0.24) return '#12FF80' // Green (14% CHANCE)
+  if (chance <= 0.17) return '#12A154' // Dark Green (3% CHANCE)
+  if (chance <= 0.2) return '#124228' // Darker Green (3% CHANCE)
+
+  return 'currentColor'
+}
```
Comment (iamacook) [inline src/components/DataRoom/CryptoPunks/utils/getRandomColor.ts:21]: This duplicates `getColor`.

File: src/components/DataRoom/TransactionsOnChain/useCounterScroll.ts [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+// Custom hook for managing counter based on scroll
+export function useCounterScroll(scrollYProgress: MotionValue<number>, percentage: number): number {
+  const [value, setValue] = useState(0)
+
+  useEffect(() => {
+    const checkScrollProgress = () => {
+      const latest = scrollYProgress.get()
+      if (latest >= 0.21) {
+        setValue(percentage)
+      } else if (latest <= 0.2) {
```
Comment (iamacook) [inline src/components/DataRoom/TransactionsOnChain/useCounterScroll.ts:13]: I'd extract these two values to constants.

File: src/components/DataRoom/TransactionsOnChain/utils/calculateYPosition.ts [hunk 10]
```diff
@@ -0,0 +1,22 @@
+// This function calculates the Y position for individual digits in the counter animation.
+// It creates a smooth rolling effect from 0 to the targetNumber.
+
+/**
+ * Calculates the Y position for individual digits in the counter animation
+ * to create a smooth rolling effect from 0 to the targetNumber.
+ *
+ * @returns {number} The calculated Y position.
+ */
```
Comment (iamacook) [inline src/components/DataRoom/TransactionsOnChain/utils/calculateYPosition.ts:9]: The comment is duplicated
```suggestion
/**
 * Calculates the Y position for individual digits in the counter animation
 * to create a smooth rolling effect from 0 to the targetNumber.
 *
 * @returns {number} The calculated Y position.
 */
```

File: src/components/DataRoom/USDCStorage/MatterCanvas.tsx [hunk 11]
```diff
@@ -0,0 +1,125 @@
+import React, { useCallback, useEffect, useRef, useState } from 'react'
+import Matter from 'matter-js'
+import { debounce } from 'lodash'
+import type { MotionValue } from 'framer-motion'
+import type { Dimensions } from './utils/types'
+import { addWallsToWorld, createWalls } from './utils/addWallsToWorld'
+import { createCoin } from './utils/createCoin'
+import { useIsMediumScreen } from '@/hooks/useMaxWidth'
+
+const { Engine, Render, Runner, World } = Matter
```
Comment (iamacook) [inline src/components/DataRoom/USDCStorage/MatterCanvas.tsx:10]: This is a bit strange. I'd suggest we destructure in the import statement.

File: src/components/DataRoom/USDCStorage/MatterCanvas.tsx [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+    (engine: Matter.Engine, dimensions: Dimensions, imgUrl: string, spawnCoins: boolean) => {
+      if (spawnCoins) {
+        const spawnInterval = setInterval(() => {
+          if (engine?.world.bodies.length >= 20) {
+            clearInterval(spawnInterval)
+            return
+          }
+          World.add(engine.world, createCoin(dimensions, imgUrl, isMobile))
+        }, 250)
+      }
```
Comment (iamacook) [inline src/components/DataRoom/USDCStorage/MatterCanvas.tsx:54]: Nit: let's add an early return if `!spawnCoins` and we are not clearing the interval on unmount.

File: src/components/DataRoom/USDCStorage/MatterCanvas.tsx [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+    const cleanup = () => {
+      if (runnerRef.current) Runner.stop(runnerRef.current)
+      if (engineRef.current) Engine.clear(engineRef.current)
+    }
+
+    cleanup()
+    createWorld()
+
+    return cleanup
+  }, [createWorld])
```
Comment (iamacook) [inline src/components/DataRoom/USDCStorage/MatterCanvas.tsx:114]: A lot of this is confusing. I'd suggest we break it down into separate hooks, move functions outside the component and add more comments.

File: src/components/DataRoom/USDCStorage/index.tsx [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+import { motion, useScroll, useTransform } from 'framer-motion'
+import MatterCanvas from './MatterCanvas'
+import { useRef } from 'react'
+import LinksWrapper from '../LinksWrapper'
+import css from './styles.module.css'
+import MotionTypography from './MotionTypography'
+import { useIsMediumScreen } from '@/hooks/useMaxWidth'
+
+// Will be replaced with the actual value in a future PR
+const USDC_PERCENTAGE = '10%'
```
Comment (iamacook) [inline src/components/DataRoom/USDCStorage/index.tsx:12]: As before, you're calculating percentages elsewhere so I'd not hardcode the percentage but the value.

File: src/components/DataRoom/VolumeTransferred/index.tsx [hunk 15]
```diff
@@ -0,0 +1,63 @@
+import { Typography } from '@mui/material'
+import { useScroll, useTransform, motion } from 'framer-motion'
+import css from './styles.module.css'
+import { useRef } from 'react'
+import type { BaseBlock } from '@/components/Home/types'
+import LinksWrapper from '../LinksWrapper'
+
+// Will be replaced with the actual value in a future PR
+const VOLUME_AMOUNT = '$611,127,712,666'
```
Comment (iamacook) [inline src/components/DataRoom/VolumeTransferred/index.tsx:9]: This will not format according to location. I'd suggest using the value and calling the formatter with that.

File: src/content/dataroom.json [hunk 16]
```diff
@@ -0,0 +1,134 @@
...
+    "component": "common/MetaTags"
+  },
+  {
+    "component": "DataRoom/Hero",
+    "title": "Welcome to Safe Data Room",
+    "text": "Explore Mind Bending Sats from Safe Universe"
+  },
+  {
+    "component": "DataRoom/USDCStorage",
+    "title": "We store... <br/>",
```
Comment (iamacook) [inline src/content/dataroom.json:15]: Do we need the line break here?

```suggestion
    "title": "We store...<br/>",
```

File: src/content/dataroom.json [hunk 17]
```diff
@@ -0,0 +1,134 @@
...
+    "title": "CryptoPunks",
+    "text": "Some text thats changing<br/>Confirming the sources of the data<br/>Any other info regarding the stats",
+    "link": {
+      "title": "Dune dashboard",
+      "href": "https://dune.com/queries/3710102/6242478"
+    }
+  },
+  {
+    "component": "DataRoom/VolumeTransferred",
+    "title": "with all time volume <br/> transferred",
```
Comment (iamacook) [inline src/content/dataroom.json:37]:
```suggestion
    "title": "with all time volume<br/>transferred",
```

File: src/content/dataroom.json [hunk 18]
```diff
@@ -0,0 +1,134 @@
...
+    "text": "and counting...",
+    "link": {
+      "title": "Dune dashboard",
+      "href": "https://dune.com/queries/3710102/6242478"
+    }
+  },
+  {
+    "component": "DataRoom/TransactionsOnChain",
+    "text": "<b>transactions on chain</b> originate from a <b>Safe</b>",
```
Comment (iamacook) [inline src/content/dataroom.json:46]:
```suggestion
    "text": "<b>transactions on chain</b> originate from a <b>Safe Account</b>",
```

File: src/content/dataroom.json [hunk 19]
```diff
@@ -0,0 +1,134 @@
...
+      "title": "Dune dashboard",
+      "href": "https://dune.com/queries/3710102/6242478"
+    }
+  },
+  {
+    "component": "DataRoom/ExternalLinksGrid",
+    "title": "Dune Boards",
+    "items": [
+      {
+        "title": "Safe on All chains",
```
Comment (iamacook) [inline src/content/dataroom.json:57]: If this is the project, "Safe" should be italicised afaiu.

```suggestion
        "title": "Safe on all chains",
```

File: src/content/dataroom.json [hunk 20]
```diff
@@ -0,0 +1,134 @@
...
+        "image": {
+          "src": "/images/dune-icon-name-logo.png",
+          "alt": "Dune Analytics logo"
+        },
+        "link": {
+          "href": "https://dune.com/safe/all"
+        }
+      },
+      {
+        "title": "Smart account landscape",
```
Comment (iamacook) [inline src/content/dataroom.json:67]:
```suggestion
        "title": "Smart Account landscape",
```

**Review by iamacook — CHANGES_REQUESTED** [review]
If he has capacity, I'd ask @usame-algan to look over the CSS as well.

File: package.json (hunk 6 above)
Comment (DiogoSoaress) [reply package.json:34]: @Malayvasa can you explain the rationale?

File: src/components/DataRoom/CryptoPunks/index.tsx (hunk 4 above)
Comment (DiogoSoaress) [reply src/components/DataRoom/CryptoPunks/index.tsx:13]: I've addressed this in https://github.com/safe-global/safe-homepage/pull/411

File: src/components/DataRoom/CryptoPunks/index.tsx (hunk 5 above)
Comment (DiogoSoaress) [reply src/components/DataRoom/CryptoPunks/index.tsx:14]: I've addressed this in https://github.com/safe-global/safe-homepage/pull/411

File: src/components/DataRoom/CryptoPunks/utils/getRandomColor.ts (hunk 8 above)
Comment (DiogoSoaress) [reply src/components/DataRoom/CryptoPunks/utils/getRandomColor.ts:21]: Good catch. I forgot to delete the other file

File: src/components/DataRoom/USDCStorage/index.tsx (hunk 14 above)
Comment (DiogoSoaress) [reply src/components/DataRoom/USDCStorage/index.tsx:12]: I've addressed this in https://github.com/safe-global/safe-homepage/pull/411

File: src/components/DataRoom/VolumeTransferred/index.tsx (hunk 15 above)
Comment (DiogoSoaress) [reply src/components/DataRoom/VolumeTransferred/index.tsx:9]: I've addressed this in https://github.com/safe-global/safe-homepage/pull/411

File: src/components/DataRoom/CryptoPunks/utils.ts (hunk 7 above)
Comment (DiogoSoaress) [reply src/components/DataRoom/CryptoPunks/utils.ts:8]: This effect including different probabilities was tuned with @liliiaorlenko . Does the array implementation make sense nevertheless?

Comment (iamacook) [reply src/components/DataRoom/CryptoPunks/utils.ts:8]: If it's a necessity then I'd say it's fine as is but the chance values seemed quite arbitrary. Using a random index will never mean you'll have to fallback to `currentColor`. This is what I was alluding to:

```suggestion
  const colors = ['#12FF80', '#12A154', '#124228']
  return colors[Math.floor((Math.random() * colors.length))]
```

File: src/components/DataRoom/USDCStorage/MatterCanvas.tsx (hunk 13 above)
Comment (DiogoSoaress) [reply src/components/DataRoom/USDCStorage/MatterCanvas.tsx]: I've broken the component down into custom hooks and functions and add comments. Let me know if it helps understanding the logic

File: src/components/DataRoom/CryptoPunks/utils.ts (hunk 7 above)
Comment (DiogoSoaress) [reply src/components/DataRoom/CryptoPunks/utils.ts]: But sometimes you don't want to color the SVG. I understand the chance values seem random but I will trust @liliiaorlenko and @Malayvasa  came to these values together. Can you confirm @Malayvasa ?

File: src/components/DataRoom/USDCStorage/utils/createMatterRender.ts [hunk 21]
```diff
@@ -0,0 +1,21 @@
...
+
+// Create the Matter.js renderer
+export const createRenderer = (
+  canvas: React.RefObject<HTMLCanvasElement>,
+  engine: Matter.Engine,
+  dimensions: Dimensions,
+  isMobile: boolean,
+) => {
+  return Render.create({
+    canvas: canvas.current!,
```
Comment (iamacook) [inline src/components/DataRoom/USDCStorage/utils/createMatterRender.ts:12]: Nit: I'd suggest throwing here instead of asserting. Although (in the current implementation) it is always defined, this may change in the future and there would be no feedback.

File: src/components/DataRoom/USDCStorage/utils/createMatterRender.ts [hunk 22]
```diff
@@ -0,0 +1,21 @@
+import { Render } from 'matter-js'
+import type { Dimensions } from '@/components/DataRoom/USDCStorage/utils/types'
+
+// Create the Matter.js renderer
+export const createRenderer = (
```
Comment (iamacook) [inline src/components/DataRoom/USDCStorage/utils/createMatterRender.ts:5]: Nit: as this is only used within `createWorld`, I'd move it there.

File: src/components/DataRoom/USDCStorage/utils/createWorld.ts [hunk 23]
```diff
@@ -0,0 +1,34 @@
...
+import { spawnCoinsInterval } from '@/components/DataRoom/USDCStorage/utils/spawnCoinsInterval'
+import type { Dimensions } from '@/components/DataRoom/USDCStorage/utils/types'
+import { Engine, Render, Runner, type World } from 'matter-js'
+import { type MutableRefObject, type RefObject } from 'react'
+
+const GRAVITY_Y_MOBILE = 1.5
+const GRAVITY_Y_DESKTOP = 1
+
+// Create the Matter.js world with the walls, renderer, and coins and start the engine
+export const createWorld = (
```
Comment (iamacook) [inline src/components/DataRoom/USDCStorage/utils/createWorld.ts:12]: Is there any cleanup we should be returning from this?

File: src/components/DataRoom/USDCStorage/utils/createWorld.ts [hunk 24]
```diff
@@ -0,0 +1,34 @@
...
+) => {
+  const engine = Engine.create()
+  engineRef.current = engine
+  worldRef.current = engine.world
+  engine.gravity.y = isMobile ? GRAVITY_Y_MOBILE : GRAVITY_Y_DESKTOP // Speeds Up Coins For Mobile
+
+  const render = createRenderer(canvas, engine, dimensions, isMobile)
+  const walls = createWalls(dimensions)
+  addWallsToWorld(engine, walls)
+  spawnCoinsInterval(engine, dimensions, imgUrl, spawnCoins, isMobile)
```
Comment (iamacook) [inline src/components/DataRoom/USDCStorage/utils/createWorld.ts:29]: We aren't handling the cleanup returned by this.

File: src/components/DataRoom/USDCStorage/utils/createWorld.ts (hunk 23 above)
Comment (iamacook) [inline src/components/DataRoom/USDCStorage/utils/createWorld.ts:12]: As above, I think you can locate these "creators" with the hook that uses them as they are only used the once. It's a bit hard to follow otherwise in separate files.

File: src/components/DataRoom/USDCStorage/utils/useResizeCanvas.ts [hunk 25]
```diff
@@ -0,0 +1,24 @@
+import type { Dimensions } from '@/components/DataRoom/USDCStorage/utils/types'
+import { debounce } from 'lodash'
+import { useEffect, useState } from 'react'
+
+// Custom Hook for handling canvas resize
+export function useResizeCanvas(isMobile: boolean) {
+  const [dimensions, setDimensions] = useState<Dimensions>({ width: 500, height: 500 })
```
Comment (iamacook) [inline src/components/DataRoom/USDCStorage/utils/useResizeCanvas.ts:7]: Just FYI: as the listener is debounced, this will remain the dimensions for the initial 300ms.

File: src/components/DataRoom/USDCStorage/utils/useMatterWorld.ts [hunk 26]
```diff
@@ -0,0 +1,33 @@
...
+export function useMatterWorld({ canvas, dimensions, isMobile, imgUrl, scrollYProgress }: PropTypes) {
+  const worldRef = useRef<World | null>(null)
+  const engineRef = useRef<Engine | null>(null)
+  const runnerRef = useRef<Runner>(null)
+  const spawnCoins = useSetSpawnCoins(scrollYProgress)
+
+  // Cleanup the world when the component unmounts
+  useEffect(() => {
+    const cleanup = () => {
+      if (runnerRef.current) Runner.stop(runnerRef.current)
```
Comment (iamacook) [inline src/components/DataRoom/USDCStorage/utils/useMatterWorld.ts:24]: `runnerRef.current` is never set so there is no cleanup happening.

File: src/components/DataRoom/USDCStorage/utils/useMatterWorld.ts [hunk 27]
```diff
@@ -0,0 +1,33 @@
...
+      if (engineRef.current) Engine.clear(engineRef.current)
+    }
+
+    cleanup()
```
Comment (iamacook) [inline src/components/DataRoom/USDCStorage/utils/useMatterWorld.ts:28]: The instantiation of the world, engine and runner is quite confusing here as they happen in nested functions. On top of moving the respective "creators" to this file (not necessarily within the hook), I'd suggest the refs instantiate their respective entities immediately as they are then not-nullable and the cleanups can be called directly, e.g.

```ts
  // Extract the following from the "creators"
  const engineRef = useRef<Engine>(Engine.create())
  const world = engineRef.current.world

  useEffect(() => {
    // Return the runner from the "creator"
    const runner = createRunner()
    
    return () => {
      Runner.stop()
      // The following is based on - https://www.fabiofranchino.com/log/matter-js-cleanup-and-destroy-instances/
      Render.stop(render)
      World.clear(engine.world)
      Engine.clear(engine)
    }
```

**Review by iamacook — APPROVED** [review]
As per discussion, the refactor I mentioned in my previous review will be addressed later.
